src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs

```diff
@@ -0,0 +1,216 @@
+        private readonly int _mantissaMask;
+        private readonly int _mantissaShift;
+
+
+        private struct Bucket
+        {
+            public Bucket(double value, int count)
+            {
+                Value = value;
+                Count = count;
+            }
+            public double Value;
+            public int Count;
+        }
+
+        public ExponentialHistogramAggregator(QuantileAggregation config)
+        {
+            _config = config;
+            _counters = new int[ExponentArraySize][];
+            if (_config.MaxRelativeError < MinRelativeError)
+            {
+                // Ensure that we don't create enormous histograms trying to get overly high precision
+                throw new ArgumentException();
+            }
+            int mantissaBits = (int)Math.Ceiling(Math.Log(1 / _config.MaxRelativeError, 2)) - 1;
+            _mantissaShift = 52 - mantissaBits;
+            _mantissaMax = 1 << mantissaBits;
+            _mantissaMask = _mantissaMax - 1;
+        }
+
+        public override IAggregationStatistics Collect()
+        {
+            int[][] counters;
+            int count;
+            lock (this)
+            {
+                counters = _counters;
+                count = _count;
+                _counters = new int[ExponentArraySize][];
+                _count = 0;
+            }
+            QuantileValue[] quantiles = new QuantileValue[_config.Quantiles.Length];
+            int nextQuantileIndex = 0;
+            if (nextQuantileIndex == _config.Quantiles.Length)
+            {
+                return new HistogramStatistics(quantiles);
+            }
+
+            // Reduce the count if there are any NaN or +/-Infinity values that were logged
+            count -= GetInvalidCount(counters);
+
+            // Consider each bucket to have N entries in it, and each entry has value GetBucketCanonicalValue().
+            // If all these entries were inserted in a sorted array, we are trying to find the value of the entry with
+            // index=target.
+            int target = QuantileToRank(_config.Quantiles[nextQuantileIndex], count);
+
+            // the total number of entries in all buckets iterated so far
+            int cur = 0;
+            foreach (Bucket b in IterateBuckets(counters))
+            {
+                cur += b.Count;
+                while (cur > target)
+                {
+                    quantiles[nextQuantileIndex] = new QuantileValue(
+                        _config.Quantiles[nextQuantileIndex], b.Value);
+                    nextQuantileIndex++;
+                    if (nextQuantileIndex == _config.Quantiles.Length)
+                    {
+                        return new HistogramStatistics(quantiles);
+                    }
+                    target = QuantileToRank(_config.Quantiles[nextQuantileIndex], count);
+                }
+            }
+
+            Debug.Assert(count == 0);
+#if NO_ARRAY_EMPTY_SUPPORT
```

Comment:
What platform do we target that doesn't have Array.Empty?

Reply:
It was one of the old ones (net45 or netstandard1.1?). I think it was recently eliminated as a target so after the rebase this became dead code.